python/ray/tests/test_autoscaler_aws.py

```diff
@@ -0,0 +1,316 @@
+            }]
+        },
+        service_response={"KeyPairs": [DEFAULT_KEY_PAIR]})
+
+
+def _configure_subnet_default_stubs(ec2_client_stub):
+    ec2_client_stub.add_response(
+        "describe_subnets",
+        expected_params={},
+        service_response={"Subnets": [DEFAULT_SUBNET]})
+
+
+def _skip_to_configure_sg_stubs(ec2_client_stub, iam_client_stub):
+    _configure_iam_role_default_stubs(iam_client_stub)
+    _configure_key_pair_default_stubs(ec2_client_stub)
+    _configure_subnet_default_stubs(ec2_client_stub)
+
+
+def _describe_subnets_echo_stub(ec2_client_stub, subnet):
+    ec2_client_stub.add_response(
+        "describe_subnets",
+        expected_params={
+            "Filters": [{
+                "Name": "subnet-id",
+                "Values": [subnet["SubnetId"]]
+            }]
+        },
+        service_response={"Subnets": [subnet]})
+
+
+def _describe_no_security_groups_stub(ec2_client_stub):
+    ec2_client_stub.add_response(
+        "describe_security_groups",
+        expected_params={"Filters": ANY},
+        service_response={})
+
+
+def _create_sg_echo_stub(ec2_client_stub, security_group):
+    ec2_client_stub.add_response(
+        "create_security_group",
+        expected_params={
+            "Description": security_group["Description"],
+            "GroupName": security_group["GroupName"],
+            "VpcId": security_group["VpcId"]
+        },
+        service_response={"GroupId": security_group["GroupId"]})
+
+
+def _describe_sgs_on_vpc_stub(ec2_client_stub, vpc_ids, security_groups):
+    ec2_client_stub.add_response(
+        "describe_security_groups",
+        expected_params={"Filters": [{
+            "Name": "vpc-id",
+            "Values": vpc_ids
+        }]},
+        service_response={"SecurityGroups": security_groups})
+
+
+def _authorize_sg_ingress_stub(ec2_client_stub, security_group):
+    ec2_client_stub.add_response(
+        "authorize_security_group_ingress",
+        expected_params={
+            "GroupId": security_group["GroupId"],
+            "IpPermissions": security_group["IpPermissions"]
+        },
+        service_response={})
+
+
+def _describe_sg_echo_stub(ec2_client_stub, security_group):
+    ec2_client_stub.add_response(
+        "describe_security_groups",
+        expected_params={"GroupIds": [security_group["GroupId"]]},
+        service_response={"SecurityGroups": [security_group]})
+
+
+CONFIG_DEFAULT_RULES_DIFFERENT_VPC = {
+    "cluster_name": "security-groups",
+    "min_workers": 1,
+    "max_workers": 1,
+    "provider": {
+        "type": "aws",
+        "region": "us-west-2",
+        "availability_zone": "us-west-2a",
+    },
+    "auth": {
+        "ssh_user": "ubuntu",
+    },
+    "head_node": {
+        "SubnetIds": [DEFAULT_SUBNET["SubnetId"]]
+    },
+    "worker_nodes": {
+        "SubnetIds": [AUX_SUBNET["SubnetId"]]
+    }
+}
+
+
+def test_create_sg_different_vpc_same_rules(iam_client_stub, ec2_client_stub):
```

Comment:
Could we remove some of this setup if we tested `_configure_security_group` directly?

Reply:
I think we could wind up with a smaller code footprint, but at the potential cost of more ambiguous upstream dependencies due to explicitly setting only the config variables that we know _configure_security_group() depends on today (like subnet IDs). Stubbing out the upstream config bootstrapping steps requires more code, but allows us to show how we logically arrive at _configure_security_group()  with a complete set of upstream dependencies.

Part of the problem here perhaps lies in the fact that we only pass the monolithic config object around between these methods, which in turn creates assumptions about which parts of the config have been populated by prior steps. While this type of design pattern makes for very stable method signatures, it also makes isolated unit testing difficult due to hiding method dependencies inside this monolithic config object.

I'd like to circle back to this in a subsequent PR focused just on this test suite. Ultimately, I think we may wind up with many more detailed tests interacting directly with lower level methods, along with a few high-level tests like this that effectively act as a stubbed integration test of the full config bootstrapping workflow.